clustered/integration-test/src/test/java/org/ehcache/clustered/BasicCacheOpsMultiThreadedTest.java

```diff
@@ -0,0 +1,208 @@
+import org.ehcache.config.units.MemoryUnit;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.terracotta.testing.rules.Cluster;
+
+import com.tc.util.Assert;
+
+import java.io.File;
+import java.io.Serializable;
+import java.net.URI;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+import static org.terracotta.testing.rules.BasicExternalClusterBuilder.newCluster;
+
+public class BasicCacheOpsMultiThreadedTest extends ClusteredTests {
+
+  private static final String RESOURCE_CONFIG =
+    "<config xmlns:ohr='http://www.terracotta.org/config/offheap-resource'>"
+    + "<ohr:offheap-resources>"
+    + "<ohr:resource name=\"primary-server-resource\" unit=\"MB\">64</ohr:resource>"
+    + "</ohr:offheap-resources>" +
+    "</config>\n";
+
+  @ClassRule
+  public static Cluster CLUSTER =
+    newCluster().in(new File("build/cluster")).withServiceFragment(RESOURCE_CONFIG).build();
+
+  @BeforeClass
+  public static void waitForActive() throws Exception {
+    CLUSTER.getClusterControl().waitForActive();
+  }
+
+  private static final String CLUSTERED_CACHE_NAME    = "clustered-cache";
+  private static final String SYN_CACHE_NAME = "syn-cache";
+  private static final String PRIMARY_SERVER_RESOURCE_NAME = "primary-server-resource";
+  private static final String CACHE_MANAGER_NAME = "/crud-cm";
+  private static final int PRIMARY_SERVER_RESOURCE_SIZE = 4; //MB
+  private static final int NUM_THREADS = 20;
+
+  private AtomicReference<Throwable> exception = new AtomicReference<>();
+
+
+  @Test
+  public void testMulipleClients() throws Throwable {
+    final CountDownLatch latch = new CountDownLatch(NUM_THREADS + 1);
+
+    List<Thread> threads = new ArrayList<>(NUM_THREADS);
+    for (int i = 0; i < NUM_THREADS; i++) {
+      Thread t1 = new Thread(content(latch));
+      t1.start();
+      threads.add(t1);
+    }
+
+    latch.countDown();
+    latch.await();
+
+    for (Thread t : threads) {
+      t.join();
+    }
+
+    Throwable throwable = exception.get();
+    if (throwable != null) {
+      throw throwable;
+    }
+  }
+
+  private Runnable content(final CountDownLatch latch) {
+    return () -> {
+      try {
+        PersistentCacheManager cacheManager = createCacheManager(CLUSTER.getConnectionURI());
+
+        latch.countDown();
+        try {
+          latch.await();
+        } catch (InterruptedException e) {
+          // continue
+        }
+
+        cacheManager.init();
+        doSyncAndPut(cacheManager);
+      } catch (Throwable t) {
+        exception.compareAndSet(null, t); // only keep the first exception
+      }
+    };
+  }
+
+  private void doSyncAndPut(PersistentCacheManager cacheManager) throws InterruptedException {
```

Comment:
Complicated question: Could you have a simpler test by adding hooks in the code to let you wait at the right places with a latch to reproduce the bug?

Reply:
I think at some point we should look at tools like byteman to do what you suggested in a clean way..

http://byteman.jboss.org/

Reply:
If you go heavyweight yes. Easier is just to put such a hook in the code.

Reply:
Ok. I will try. But I am personally not a big fan of adding test specific code in production code.. If we can mock things and achieve the same it is OK..but in this case, due to the multi-threaded nature of the problem, it is very hard to achieve the same result using mocking. Let me give it a try though.

Reply:
@henri-tremblay created a unit test using mocks..Managed to do it without stubs in production code. Pls take a look.

Reply:
All good (just a little comment)